topics/data-science/tutorials/python-flow/tutorial.md

```diff
@@ -0,0 +1,476 @@
+
+The following Python code uses `elif` to print the sign of a number.
+
+```python
+num = -3
+
+if num > 0:
+    print(f'{num} is positive')
+elif num == 0:
+    print(f'{num} is zero')
+else:
+    print(f'{num} is negative')
+```
+
+**NB**: To test for equality we use a double equals sign `==`
+rather than a single equals sign `=` which is used to assign values.
+
+## Combining Tests
+
+We can also combine tests using `and` and `or`.
+`and` is only true if both parts are true:
+
+```python
+a = 1
+b = -1
+if (a > 0) and (b <= 0):
+    print('both parts are true')
+else:
+    print('at least one part is false')
+```
+
+> ### {% icon question %} Question: Predict what happens
+> Predict the outcomes of the following values of `a` and `b` above. Predicting what you think the code will do is a useful skill to practice
+> 1. a = 0; b = -1
+> 2. a = 0; b = 10
+> 3. a = 4; b = -22
+> 4. a = 99; b = 99
+>
+> > ### {% icon solution %} Solution
+> > 1. at least one part is false
+> > 2. at least one part is false
+> > 3. both parts are true
+> > 4. at least one part is false
+> {: .solution}
+{: .question}
+
+
+while `or` is true if at least one part is true:
+
+```python
+a = 1
+b = -1
+if (a < 0) or (b > 0):
+    print('at least one test is true')
+```
+
+
+> ### {% icon tip %} Tip: `True` and `False`
+> `True` and `False` are special words in Python called `booleans`,
+> which represent truth values. A statement such as `1 < 0` returns
+> the value `False`, while `-1 < 0` returns the value `True`.
+{: .tip}
+
+`True` and `False` booleans are not the only values in Python that are true and false.
+In fact, *any* value can be used in an `if` or `elif`. This is commonly used to
+check, for instance, if a string is empty or if some data is provided:
+
+```python
+if '':
+    print('empty string is true')
+if 'word':
+    print('word is true')
+```
+
+You can also use it to check if a list is empty or full:
+
+```python
+if []:
+    print('empty list is true')
+if [1, 2, 3]:
+    print('non-empty list is true')
+# The last statement is equivalent to:
+if len([1, 2, 3]) > 0:
+    print('non-empty list is true')
+```
+
+Or you can check if a number is zero, or non-zero:
+
+```python
+if 0:
+    print('zero is true')
+if 1:
+    print('one is true')
+```
+
+## Inverting Conditions
+
+Sometimes it is useful to check whether some condition is not true.
+The Boolean operator `not` can do this explicitly.
+After reading and running the code below,
+write some `if` statements that use `not` to test the rule
+that you formulated in the previous question.
+`not` is a `unary` (not `binary`) operator: it only takes a single value
+
+```python
+if not '':
+    print('empty string is not true')
+if not 'word':
+    print('word is not true')
+if not not True:
+    print('not not True is true')
+```
+
+## Ranges
+
+Python makes it super easy to check if a number is within a range.
+
+```python
+# quality_score = calculate_score(my_fastq_file)
+
+if quality_score > 40:
```

Comment:
I LOVE THIS!

But is it the goal that the funcion is commented out and it runs into an error?
or can we put there a dummy value

Reply:
Good idea, adding a dummy.